### PR TITLE
[dev] (feat): Native implementation for Attention Residuals (AttnRes) with Block AttnRes, pipeline parallelism and MTP support

### PR DESCRIPTION
**Pull request by jingqiny-99**
# Add Attention Residuals (AttnRes) with Block AttnRes, pipeline parallelism and MTP support

## What

Implements **Attention Residuals** (Kimi Team, [arXiv:2603.15031](https://arxiv.org/abs/2603.15031)): the fixed PreNorm residual accumulation is replaced by a per-token softmax attention over depth sources — the token embedding, completed depth-block sums, and the running intra-block partial sum. Each self-attention and MLP sublayer aggregates the sources with its own zero-initialized pseudo-query and RMSNorm key weight; the final output head (and each MTP depth) aggregates all sources before the final layernorm.

New config surface:

- `--enable-attention-residuals`
- `--attn-res-block-layers N` — transformer layers per depth block (block size S = 2N sublayers). Total depth sources = `floor((num_layers-1)/N) + 2`; the paper's sweet spot is ~8–10 sources.

## Design

- **Aggregation module** (`attention_residual.py`): `h = Σ_j softmax_j(w·RMSNorm(V_j)) · V_j` as a custom autograd Function. Statistics/softmax/accumulation run in fp32; only value references plus per-token fp32 scalars (dots, inverse RMS, softmax weights) are saved, and keys/logits are recomputed in backward — a naive autograd graph would retain O(n) fp32 `[s,b,h]` intermediates per sublayer. Output is cast back to the residual dtype, so fp32 never leaks into the stream. `w` is zero-initialized (uniform initial weights, required for stability per the paper); with RMSNorm scale invariance this makes the network functionally identical to the baseline at initialization, which the tests exploit as an equivalence oracle.
- **Layer** (`AttnResTransformerLayer`): the running state carried between layers is the intra-block partial sum; block boundaries and source arity are static per-layer constants. The depth-source tuple is threaded per call by `TransformerBlock` — entries are references to activations that are already alive, so the bookkeeping adds no copies within a stage. Reuses the base `_forward_self_attention_output_with_bias` / `_forward_mlp_output_with_bias` / `_forward_post_mlp` helpers, so MoE (incl. shared-expert overlap and the packed-sequence unflatten), selective recompute and dtype paths are inherited rather than duplicated.
- **Pipeline parallelism** (non-interleaved): sources + partial cross stage boundaries as one tensor concatenated along the sequence dimension; per-boundary slice counts are derived statically from the pipeline layer layout in `get_tensor_shapes` (incl. `account_for_embedding/loss_in_pipeline_split`). RoPE length derivation on non-first stages divides the payload slice count back out.
- **MTP**: each depth's transformer layer attends over the trunk depth history with its own partial sum (the eh_proj output), and applies a per-depth final aggregation before its final layernorm. `mtp_detach_heads` also detaches the depth sources.
- **Aggregation math is per-token and channel-local**: transparent to TP (hidden dim intact under SP), CP, and EP; the only parallelism dimension with new communication is PP.

## Supported / WIP

Supported in this first slice: eager bf16/fp8 training, TP/SP/CP/EP/MoE, non-interleaved PP, selective recompute (`core_attn`, `layernorm`, `mlp`, `moe`), MTP in the standard last-stage placement, THD/packed sequences, distributed checkpointing.

WIP: interleaved VPP, CUDA graphs, full recompute, the EP-overlap fine-grained schedule, CPU/fine-grained offloading, `fused_residual_rmsnorm`, `fp32_residual_connection`, custom pipeline layouts, hybrid stacks, inference, feature extraction.

## Testing

- `tests/unit_tests/transformer/test_attention_residual.py`:
  - forward/backward of the aggregation vs a plain-autograd fp64 reference (1–10 sources, bf16, THD layouts);
  - zero-init uniformity + nonzero pseudo-query gradient at init;
  - block/boundary schedule helpers and payload pack/unpack round-trips (grad flow through slice views);
  - config validation walls;
  - **init-equivalence oracle**: with tiny norm eps, a 4-layer GPT with AttnRes matches the baseline logits at initialization (this catches wiring/boundary/dtype mistakes in one shot; note the equivalence is exact only as eps→0 — production eps values break RMSNorm scale invariance for small activations).
- Local CPU pipeline check: the same equivalence oracle run through the real non-interleaved schedule on 2 gloo ranks (PP2), max logit diff 1.4e-6 vs baseline, exercising `get_tensor_shapes` widths, payload pack/unpack and cross-stage RoPE length.
- GPU verification on GB200 (DSv3-like 16L MoE proxy, PP2/EP2 + MTP, bf16, TE) is in progress; results will be posted on this PR.

## Notes for review

- Block boundaries fall on transformer-layer starts only (S is always an even number of sublayers); Full AttnRes (S=1 sublayer) and odd S are intentionally out of scope for this slice.
- At zero init the key-norm weight receives zero gradient until the pseudo-query moves (chain rule through `q = w⊙γ`); the factorization is redundant at w=0, so no expressivity is lost.
- Known follow-up for PP performance: the payload grows with completed blocks (~(n+2)× the base hidden per transition late in the pipeline), so non-interleaved PP is correctness-complete but comm-heavy; interleaved VPP + incremental cross-stage caching (paper §4.1) is the planned perf-bearing configuration.

### CODE REVIEW

**Comment by jingqiny-99** [comment]
First GPU verification round (GB200 x4, single node, bf16 + TE, DSv3-like 16L MoE proxy, PP2/EP2/SP, MTP depth 1, `--attn-res-block-layers 2`, mock data, 100 iters):

- iter-1 lm loss matches the baseline arm to 4 decimals (11.9702 vs 11.9704) — the zero-init equivalence holds on the production stack across a real PP boundary, so the payload/boundary/final-aggregation wiring checks out end to end.
- trains cleanly, no NaN; final val loss 2.0e-2. Loss drops consistently faster than the baseline arm on this (memorization-flavored) mock task, e.g. 1.80 vs 5.09 at iter 50.
- MTP branch works with the depth-attention inside (mtp_1 loss tracks lm loss from ~11.97).
- cost of the unfused eager aggregation at this tiny proxy scale: 62.9 vs 91.6 TFLOP/s/GPU median, peak mem +~270MB — fused kernel is the planned follow-up, this slice is about correctness.

btw both arms ran on the same commit (d20301b), baseline = same config minus the two flags.

**Comment by yuzhongw-nvidia** [comment]
Hi @jingqiny-99 , could you please also support it for hybrid model? Thanks.

**Comment by jingqiny-99** [comment]
> Hi @jingqiny-99 , could you please also support it for hybrid model? Thanks.

OK, I'll support the hybrid model soon.

File: megatron/core/transformer/attention_residual.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    alpha = torch.softmax(dots * rstds, dim=0)  # [n, ...] fp32
+
+    out32 = None
+    for j, value in enumerate(values):
+        term = alpha[j].unsqueeze(-1) * value.float()
+        out32 = term if out32 is None else out32 + term
+    return out32.to(values[0].dtype), alpha, dots, rstds
+
+
+def _attn_res_bwd_math(pseudo_query, key_norm_weight, alpha, dots, rstds, grad_output, values):
```
Comment (Wohox) [inline megatron/core/transformer/attention_residual.py:599]: Why not use pytorch's autograd to build the backward computation graph automatically?
If this is to support `torch.compile` backward, I would recommend to skip this function when not enabling torch.compile backend.

FLA open sourced a fused kernel for AttnRes training, maybe we can evaluate it against torch.compile.

**Comment by Wohox** [comment]
Do we have plan for AttnResidual VPP cache, this PR currently doesn't support it.

File: megatron/core/transformer/attention_residual.py (hunk 1 above)
Comment (jingqiny-99) [reply megatron/core/transformer/attention_residual.py:599]: > FLA open sourced a fused kernel for AttnRes training, maybe we can evaluate it against torch.compile.

Agree, I'll test it and try to integrate in this PR later.

**Comment by jingqiny-99** [comment]
> Do we have plan for AttnResidual VPP cache, this PR currently doesn't support it.

I think it now supports VPP, but maybe need more tests.

File: megatron/core/transformer/transformer_config.py [hunk 2]
```diff
@@ -1330,6 +1332,44 @@ class TransformerConfig(ModelParallelConfig):
...
+
+    attn_res_impl: str = "eager"
+    """Implementation of the AttnRes depth aggregation: 'eager' (a memory-lean custom autograd
+    Function built from plain PyTorch ops) or 'compile' (a plain PyTorch forward wrapped in
+    torch.compile, with AOTAutograd generating its backward and one specialization per depth
+    arity; falls back to the eager custom Function with a warning if compilation is unavailable),
+    or 'fla' (FLA's three-kernel fused training implementation with checkpoint_level=1; requires
+    flash-linear-attention). The eager loop is CPU-dispatch-bound — measured ~3-4 ms of CPU wall
+    per aggregation on GB200 at small hidden sizes — so 'fla' is recommended when the optional
+    dependency is installed, with 'compile' as the dependency-free optimized path."""
```
Comment (yuzhongw-nvidia) [inline megatron/core/transformer/transformer_config.py]:
```suggestion
    attn_res_impl: str = "fla"
    """Implementation of the AttnRes depth aggregation.
    Options:
    - 'eager': a memory-lean custom autograd Function built from plain PyTorch ops.
    - 'compile': a plain PyTorch forward wrapped in torch.compile, with AOTAutograd
      generating its backward and one specialization per depth arity; falls back to the
      eager custom Function with a warning if compilation is unavailable.
    - 'fla': FLA's three-kernel fused training implementation with checkpoint_level=1;
      requires flash-linear-attention.
    The eager loop is CPU-dispatch-bound — measured ~3-4 ms of CPU wall
    per aggregation on GB200 at small hidden sizes — so 'fla' is recommended when the optional
    dependency is installed, with 'compile' as the dependency-free optimized path."""
```
